src/Workspaces/Core/Portable/Log/HistogramLogAggregator.cs

```diff
@@ -1,53 +1,95 @@
+    internal sealed class HistogramLogAggregator : AbstractLogAggregator<HistogramLogAggregator.HistogramCounter>
     {
         private readonly int _bucketSize;
         private readonly int _maxBucketValue;
+        private readonly int _bucketCount;
 
         public HistogramLogAggregator(int bucketSize, int maxBucketValue)
         {
+            if (bucketSize <= 0 || maxBucketValue <= 0 || maxBucketValue % bucketSize != 0)
+            {
+                throw new ArgumentException();
+            }
+
             _bucketSize = bucketSize;
             _maxBucketValue = maxBucketValue;
+            _bucketCount = maxBucketValue / bucketSize + 1;
         }
 
-        public void IncreaseCount(decimal value)
+        protected override HistogramCounter CreateCounter()
         {
-            var counter = GetCounter(GetBucket(value));
-            counter.IncreaseCount();
+            return new HistogramCounter(_bucketSize, _maxBucketValue, _bucketCount);
         }
 
-        public int GetCount(decimal value)
+        public void IncreaseCount(object key, decimal value)
         {
-            if (TryGetCounter(GetBucket(value), out var counter))
-            {
-                return counter.GetCount();
-            }
-
-            return 0;
+            var counter = GetCounter(key);
+            counter.IncreaseCount(value);
         }
 
-        protected override LogAggregator.Counter CreateCounter()
+        internal sealed class HistogramCounter
         {
-            return new LogAggregator.Counter();
-        }
+            private readonly int[] _buckets;
 
-        private int GetBucket(decimal value)
-        {
-            var bucket = (int)Math.Floor(value / _bucketSize) * _bucketSize;
-            if (bucket > _maxBucketValue)
+            public int BucketCount { get; }
+            public int BucketSize { get; }
+            public int MaxBucketValue { get; }
+
+            public HistogramCounter(int bucketSize, int maxBucketValue, int bucketCount)
+            {
+                Debug.Assert(bucketSize > 0 && maxBucketValue > 0 && bucketCount > 0);
+
+                BucketSize = bucketSize;
+                MaxBucketValue = maxBucketValue;
+                BucketCount = bucketCount;
+                _buckets = new int[BucketCount];
+            }
+
+            public void IncreaseCount(decimal value)
             {
-                bucket = _maxBucketValue;
+                var bucket = GetBucket(value);
+                _buckets[bucket]++;
             }
 
-            return bucket;
+            public string GetBucketsAsString()
+            {
+                var pooledStringBuilder = PooledStringBuilder.GetInstance();
+                var builder = pooledStringBuilder.Builder;
+
+                builder.Append('[');
+                builder.Append(_buckets[0]);
+
+                for (var i = 1; i < _buckets.Length; ++i)
```

Comment:
just curious, not using `string.Join` to avoid allocations?

2nd question - does it matter a lot here?  this should only run once when shutdown occurs, or does it run more often for the completion use case?

Reply:
1. Yup, this is just to avoid allocation.
2. Probably not much, I went with this because the code is still quite simple.